### PR TITLE
[upload] Update upload subsystem for the new Red Hat upload API

### PR DESCRIPTION
**Pull request by jcastill**
The Red Hat Customer Portal is migrating to a new upload API, so the upload target needs to be updated to use the new endpoints and flow.

Archives up to 5 GB are uploaded in a single HTTP PUT request, while archives exceeding that threshold are split into 128 MiB chunks and uploaded in parallel. As a consequence of the new upload limits, we don't perform fallback to SFTP depending on upload size, only when the upload fails or the user selects it explicitly.

The multipart flow handles session initialization, per-chunk retry, session expiration checks, and automatic abort on failure. A new --upload-threads option (1-16, default 4) controls the number of parallel upload threads per chunk used.

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X] Is the subject and message clear and concise?
- [X] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [X] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [X] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?
- [X] Are all passwords or private data gathered by this PR [obfuscated](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin#how-to-prevent-collecting-passwords)?

### CODE REVIEW

**Comment by jcastill** [comment]
Some samples of the PR execution. Please excuse the wall of text, but I think it's important to show how the upload will look like, specially since the tests we can run are only in an stage environment and we have very limited options to test this and get stage accounts:

* Upload of big file (more than 5GB) without verbosity:

```
$ sudo bin/sos upload --case-id 05118085 ../upload_files/testfile_5GB_plus1.bin --upload-target=redhat

sos upload (version 4.11.2)
This utility is used to upload files to a target location based either on a
command line option or detecting the local distribution.

The archive to be uploaded may contain data considered sensitive and its content
should be reviewed by the originating organization before being passed to any
third party.

No configuration changes will be made to the system running this utility.

Press ENTER to continue, or CTRL-C to quit

Attempting upload to Red Hat Customer Portal
Please visit the following URL to authenticate this device: https://sso.stage.redhat.com/device?user_code=BSLD-IPMP
Device authorized correctly. Uploading file to Red Hat Customer Portal
Upload will split the 5.0G report into multiple parts and send 4 chunk(s) in parallel
Starting multipart upload: 41 chunks with 4 threads.
Progress: 1/41 chunks uploaded
Progress: 2/41 chunks uploaded
[...]
Progress: 40/41 chunks uploaded
Progress: 41/41 chunks uploaded
Upload completed successfully.
File ../upload_files/testfile_5GB_plus1.bin uploaded successfully
```


* Upload of file bigger than limit with verbosity:

```
$ sudo bin/sos upload --case-id 05118085 ../upload_files/testfile_5GB_plus1.bin --upload-target=redhat -vvv
[sudo] password for jcastill: 

sos upload (version 4.11.2)
This utility is used to upload files to a target location based either on a
command line option or detecting the local distribution.

The archive to be uploaded may contain data considered sensitive and its content
should be reviewed by the originating organization before being passed to any
third party.

No configuration changes will be made to the system running this utility.

Press ENTER to continue, or CTRL-C to quit

Size of archive is bigger than Red Hat Customer Portal limit for simple uploads of 5.0G. Using multi-part upload.

Attempting upload to Red Hat Customer Portal
Please visit the following URL to authenticate this device: https://sso.stage.redhat.com/device?user_code=ETRG-UWVB
The SSO authentication is successful
Device authorized correctly. Uploading file to Red Hat Customer Portal
Upload strategy: multipart
Upload will split the 5.0G report into multiple parts and send 4 chunk(s) in parallel
Multipart upload parameters: archive_size=5368709121, chunk_size=134217728, total_chunks=41
Requesting upload session: file=testfile_5GB_plus1.bin, size=5368709121, chunks=41
Upload session response: 200
Upload session initialized: attachmentId=aff911d44d674cc380f47ba4a0d-msn5dctb, strategy=MULTIPART, parts=41
Starting multipart upload: 41 chunks with 4 threads.
Part 2 uploaded successfully
Progress: 1/41 chunks uploaded
Part 3 uploaded successfully
Progress: 2/41 chunks uploaded
Part 1 uploaded successfully
Progress: 3/41 chunks uploaded
Part 4 uploaded successfully
Progress: 4/41 chunks uploaded
Part 5 uploaded successfully
Progress: 5/41 chunks uploaded
Part 6 uploaded successfully
Progress: 6/41 chunks uploaded
Part 8 uploaded successfully
Progress: 7/41 chunks uploaded
[...]
Part 39 uploaded successfully
Progress: 40/41 chunks uploaded
Part 40 uploaded successfully
Progress: 41/41 chunks uploaded
Completing upload: attachmentId=aff911d44d674cc380f47ba4a0d-msn5dctb, parts=41
Upload completed successfully.
File ../upload_files/testfile_5GB_plus1.bin uploaded successfully
```

* Single HTTP upload (files less than 5Gb):

```
$ sudo bin/sos upload --case-id 05118085 ../upload_files/testfile_10MB.bin --upload-target=redhat 
[sudo] password for jcastill: 

sos upload (version 4.11.2)
This utility is used to upload files to a target location based either on a
command line option or detecting the local distribution.

The archive to be uploaded may contain data considered sensitive and its content
should be reviewed by the originating organization before being passed to any
third party.

No configuration changes will be made to the system running this utility.

Press ENTER to continue, or CTRL-C to quit

Attempting upload to Red Hat Customer Portal
Please visit the following URL to authenticate this device: https://sso.stage.redhat.com/device?user_code=QWAP-IWQT
Device authorized correctly. Uploading file to Red Hat Customer Portal
Uploading 10.0M report in a single request
Response status: 200
File ../upload_files/testfile_10MB.bin uploaded successfully
```

* Single HTTP upload (files less than 5Gb) with verbosity:

```
sos upload (version 4.11.2)
This utility is used to upload files to a target location based either on a
command line option or detecting the local distribution.

The archive to be uploaded may contain data considered sensitive and its content
should be reviewed by the originating organization before being passed to any
third party.

No configuration changes will be made to the system running this utility.

Press ENTER to continue, or CTRL-C to quit

sudo bin/sos upload --case-id 05118085 ../upload_files/testfile_10MB.bin --upload-target=redhat -vvv
Attempting upload to Red Hat Customer Portal
Please visit the following URL to authenticate this device: https://sso.stage.redhat.com/device?user_code=YWQB-HPIB
The SSO authentication is successful
Device authorized correctly. Uploading file to Red Hat Customer Portal
Upload strategy: simple
Uploading 10.0M report in a single request
Simple upload: file=testfile_10MB.bin, size=10485760
Requesting upload session: file=testfile_10MB.bin, size=10485760, chunks=0
Upload session response: 200
Upload session initialized: attachmentId=a4b96d24deea4a31a0b64750643-msn7j8rn, strategy=SINGLE_PUT, parts=0
Uploading to URL: https://hydra-sct-attachments-stage.s3.us-west-2.amazonaws.com/***
Response status: 200
File ../upload_files/testfile_10MB.bin uploaded successfully
```

File: sos/upload/targets/redhat.py [hunk 1]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+        # }
+        attachment_id = self._upload_session_details.get('attachmentId')
+        upload_id = self._upload_session_details.get('uploadId')
+        presigned_parts = self._upload_session_details.get('parts', [])
+
+        if len(presigned_parts) != total_chunks:
+            self.ui_log.error(
+                f"Presigned URL count mismatch: expected {total_chunks}, "
+                f"got {len(presigned_parts)}")
+            self._abort_upload(self._build_abort_data())
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: You always call `self._abort_upload(self._build_abort_data())` with the same argument. To pass less data, does it make sense to have the `_abort_upload` method without an argument, and have there directly

```
json=self._build_abort_data(),
```

instead of current

```
json=abort_data,
```

? (well, also debug message at the beginning would have to be updated accordingly).

Self devil's advocate: this approach is coherent with complete_upload method where the info about upload is specific in each case. So wither implementation (current or my proposal) are similarly "valid", in my eyes.

File: sos/upload/targets/redhat.py [hunk 2]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+                    "An error occurred while attempting to cancel upload: "
+                    f"{response_abort.status_code}")
+        except requests.exceptions.RequestException as e:
+            self.ui_log.error(f"Error cancelling upload: {e}")
+
+    def _complete_upload(self, complete_data):
+        """Complete an upload session
+
+        :param complete_data: Information about the upload session
+        :returns: Response object onsuccess, None on failure
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: nitpick: `s/onsuccess/on success/`.

File: sos/upload/targets/redhat.py [hunk 3]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+                    f"({response_complete.status_code}): "
+                    f"{response_complete.text}")
+            return None
+        except requests.exceptions.RequestException as e:
+            self.ui_log.error("An exception occurred while "
+                              f"completing upload: {e}")
+            return None
+
+    def _get_upload_session_details(self, archive, archive_size, total_chunks):
+        """Get the details of the upload session lke attachmentId,
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: nitpick: `s/lke/like/`

File: sos/upload/targets/redhat.py [hunk 4]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+                    "Upload completion failed "
+                    f"({response_complete.status_code}): "
+                    f"{response_complete.text}")
+            return None
+        except requests.exceptions.RequestException as e:
+            self.ui_log.error("An exception occurred while "
+                              f"completing upload: {e}")
+            return None
+
+    def _get_upload_session_details(self, archive, archive_size, total_chunks):
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py:189]: Worth renaming the method to e.g. `_generate_upload_session_details`? Since `get` method should return the generated object, while we dont return anything here.

Or maybe `_init_upload_session` or something similar, as we initialize the upload here?

File: sos/upload/targets/redhat.py [hunk 5]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+        Stores the result in self._upload_session_details.
+
+        :param archive: The file object for the archive
+        :param archive_size: Size of the archive in bytes
+        :param total_chunks: Number of chunks for multipart (0 for simple)
+        """
+        # Lets reset session details before each attempt so that stale or
+        # expired presigned URLs from a previous call are not reused
+        # if this request fails and upload_archive retries.
+        self._upload_session_details = {}
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: great explanation, thanks. I was about to ask why we cant "cache" the info.

File: sos/upload/targets/redhat.py [hunk 6]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+        self.ui_log.debug(
+            f"Requesting upload session: "
+            f"file={data['fileName']}, "
+            f"size={data['fileSize']}, "
+            f"chunks={data['totalChunks']}")
+        try:
+            headers = self._get_upload_https_auth()
+            upload_url = (
+                f"{self.RH_API_HOST}"
+                f"{self.RH_HYDRA_ATTACHMENTS_PATH}/upload")
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: worth be coherent with other `requests.*` invocations and have `base = ..` and `upload_url = ..`?

(maybe we can save the `base` into the class instance at init, as `self.base`, to prevent code duplication? it has pros and cons, though..)

File: sos/upload/targets/redhat.py [hunk 7]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+
+        :param file_path: The path to the file to upload
+        :param chunk_info: Information about the chunk we are uploading
+        :param presigned_url: Corresponding presigned URL to use for this chunk
+        :param max_retries: Max number of retries in case of error
+        """
+        part_number = chunk_info["part_number"]
+        size = chunk_info["size"]
+        offset = chunk_info["offset"]
+        chunk_timeout = max(TIMEOUT_DEFAULT, size // (512 * 1024))
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py:282]: What is the `size // (512 * 1024)`? some safe assumption of a very slow upload time (512KiB per second?)? Maybe worth commenting this..?

File: sos/upload/targets/redhat.py [hunk 8]
```diff
@@ -233,18 +754,19 @@ def check_file_too_big(self, archive):
         # so we don't need to call any size converter implemented
         # in tools.py
         if size >= self._max_size_request:
-            self.ui_log.warning(
+            self.ui_log.debug(
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py:742]: subjective opinion: this is worth of warning or at least info level, since we alter the default behaviour and start doing something concurrently (with its higher memory footprint as well).

File: man/en/sos-report.1 [hunk 9]
```diff
@@ -497,6 +498,11 @@ which is typically used as a fallback target.
 
 Valid values for PROTO are: 'auto' (default), 'https', 'ftp', 'sftp'.
 .TP
+.B \--upload-threads THREADS
+Number of threads to use for multipart uploads. When an archive exceeds a
+certain size, defined in the targets, it is uploaded in parallel chunks
+using this many threads. Default is 4.
+.TP
```
Comment (pmoravec) [inline man/en/sos-report.1:531]: Worth explaining here the:

- boundaries (1-17)
- effect (higher value means faster upload, but consumes more bandwidth and more memory (128MB per thread - but dont we copy the data in memory 2 times? isnt it 256ish MB? I would need to experiment but dont have means for it..)

(then we can drop the mention of memory footprint from cmdline help, maybe?)

File: man/en/sos-upload.1 [hunk 10]
```diff
@@ -112,6 +113,12 @@ are: redhat, canonical, generic, local.
 If this option is not called, sos will try to determine the local target
 and use it for uploads.
 
+.TP
+.B \--upload-threads THREADS
```
Comment (pmoravec) [inline man/en/sos-upload.1:118]: Ditto.

File: sos/upload/targets/redhat.py [hunk 11]
```diff
@@ -118,9 +617,31 @@ def _upload_https_post(self, archive, verify=True):
...
+            )
+            response = self.multipart_upload(archive=files)
+        else:
+            self.ui_log.info(
+                f"Uploading {convert_bytes(archive_size)} report in a "
+                f"single request"
+            )
+            response = self.simple_http_upload(archive=files, verify=verify)
+        if response is None:
+            raise Exception(
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py:625]: Can't we raise some less generic exception class? Or pass the response details in the exception text?

File: sos/upload/targets/redhat.py [hunk 12]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+                f"{parsed_url.hostname}/***")
+            # TODO: for simple http uploads with files that are multiple
+            # GBs, but not quite enough to be a multipart upload, the
+            # default timeout may be too short. Revisit in the future.
+            try:
+                response = requests.put(presigned_url, data=file_obj,
+                                        headers={'Content-Type':
+                                                 'application/octet-stream'},
+                                        verify=verify,
+                                        timeout=TIMEOUT_DEFAULT)
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: Uploading a 4.9GB file can timeout on slow connections (5 GB / 300s = 17 MB/s minimum). Worth using similar formula like on another place (`max(TIMEOUT_DEFAULT, archive_size // (512 * 1024))`)?

File: sos/upload/targets/redhat.py [hunk 13]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+    def simple_http_upload(self, archive, verify=True):
+        """Simple request http upload. This is used for files that
+        have a size less than _max_size_request.
+
+        We don't need to call _complete_upload in this case:
+        the server auto-completes simple (non-multipart) uploads.
+        The complete/abort lifecycle only applies to multipart
+        uploads for files exceeding _max_size_request.
+
+        :param archive: Dict containing the file tuple under the 'file' key
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: Why dict and not the value itself (or rather it's [1] file_obj directly)?

File: sos/upload/targets/redhat.py [hunk 14]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+        self.ui_log.error(f"Upload of chunk {part_number} failed after "
+                          f"{max_retries} attempts")
+        return {"partNumber": part_number,
+                "ETag": etag,
+                "success": False}
+
+    def multipart_upload(self, archive):
+        """Manage the multipart upload for files exceeding _max_size_request.
+
+        :param archive: Dict containing the file tuple under the 'file' key
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: Why dict and not the value itself (or rather it's [1] file_obj directly)?

File: sos/upload/targets/redhat.py [hunk 15]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+        file_obj = file_tuple[1]
+        file_path = file_obj.name
+
+        thread_count = getattr(
+            self.commons.get('cmdlineopts', None),
+            'upload_threads', self._upload_thread_count)
+        self._upload_thread_count = thread_count
+
+        # Calculate number of chunks for the multipart upload
+        chunk_size = 128 * 1024 * 1024  # 128 MiB
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: Just a thought: worth moving this constatnt close to `_max_size_request` declaration?

File: sos/upload/targets/redhat.py [hunk 16]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+        archive_size = os.path.getsize(file_path)
+        total_chunks = (archive_size + chunk_size - 1) // chunk_size
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: The `+ chunk_size - 1` part means you round-divide **to up**, not down (like // does), right? Maybe worth commenting / explaining it?

File: sos/upload/targets/redhat.py [hunk 17]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+                            f"{total_chunks} chunks "
+                            "uploaded")
+                    else:
+                        failed_chunks.append(chunk_info)
+                except Exception as e:
+                    self.ui_log.error(
+                        f"Unexpected error while uploading chunk: {e}")
+                    failed_chunks.append(chunk_info)
+
+                if expiration and datetime.now(timezone.utc) >= expiration:
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py]: There is a race condition:

- some chunk gets uploaded just before expiration, still on time,
- we get successful result,
- *then* the expiration time happens
- then we evaluate this check and claim the upload failed

(if some chunk was still pending, that is correct behaviour. if it was the latest chunk we wrongly claim it failed)

Not sure how better test this, though.

File: sos/upload/targets/redhat.py [hunk 18]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+        # Lets reset session details before each attempt so that stale or
+        # expired presigned URLs from a previous call are not reused
+        # if this request fails and upload_archive retries.
+        self._upload_session_details = {}
+
+        data = {
+            'caseNumber': self.commons['cmdlineopts'].case_id,
+            'fileName': archive.name.split('/')[-1],
+            'fileSize': archive_size,
+            'totalChunks': total_chunks,
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py:207]: Just to clarify to me: for standard upload (via `simple_http_upload`), `total_chunks=0`. So the API specification distinguishes "chunkc or not chunks" based on `totalChunks != 0`. Do I get this right?

File: sos/upload/targets/redhat.py [hunk 19]
```diff
@@ -118,9 +617,31 @@ def _upload_https_post(self, archive, verify=True):
         self._device_token = RHELAuth.get_access_token()
         self.ui_log.info("Device authorized correctly. Uploading file to "
                          f"{self.get_upload_url_string()}")
-        return requests.post(self.get_upload_url(), files=files,
-                             headers=self._get_upload_https_auth(),
-                             verify=verify, timeout=TIMEOUT_DEFAULT)
+
+        strategy = 'multipart' if self._upload_multipart \
```
Comment (pmoravec) [inline sos/upload/targets/redhat.py:605]: Example

```
"uploadStrategy": "MULTIPART",
```

on line 388 has the string in capitals. What is the real value, then? Shouldnt it be unified?

**Comment by pmoravec** [comment]
I am mostly OK with the PR, thanks for the lengthy work.

Many comments I raised are rather subjective opinion or "what about" ideas of improvement - I dont insist on most of them.

I feel there are a few places here and there, where the code might get improved to be more robust, but I dont want to be too paranoic and polute the PR by my fears. Rather the daily use will reveal them, if there is any real wekaness.

File: sos/upload/targets/redhat.py (hunk 2 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py]: ... sorry. Will change this

File: sos/upload/targets/redhat.py (hunk 3 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py]: ... I'm sorry

File: sos/upload/targets/redhat.py (hunk 4 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py:189]: The idea behind this name is that on the server side we get a new, different upload session each time, so we are actually getting the specific session details for this upload each time. But, you know... I may have translated this in my head directly from Spanish and it doesn't make sense in English. So in that case, perhaps _init_upload_session is better?

File: sos/upload/targets/redhat.py (hunk 5 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py]: Luckily we can test in stage and I could catch this sooner

File: sos/upload/targets/redhat.py (hunk 6 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py]: what cons do you see? I only see pros tbh

File: sos/upload/targets/redhat.py (hunk 8 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py:742]: So I did it this way because I thought I was being very verbose (or the same as usual in me). When we do the upload, you have either:

`Uploading 10.0M report in a single request
`
or:

`Upload will split the 5.0G report into multiple parts and send 4 chunk(s) in parallel
`
And I thought this was more worth of a debug session, but can definitely change it back to warning.

File: man/en/sos-report.1 (hunk 9 above)
Comment (jcastill) [reply man/en/sos-report.1:531]: Ack, good point. I can add a vague note about the memory usage, and can try to create a simple python script to test memory usage. But I want to make sure that the users know the trade offs

File: sos/upload/targets/redhat.py (hunk 12 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py]: ok. I'll reword the TODO as well because I think it's worth keeping an eye on the timeout here.
A quick note, I asked if we can use multipart for less than 5Gb or reduce the limit size and was told that no, so we are stuck here with a big request

File: sos/upload/targets/redhat.py (hunk 15 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py]: ok, will make it more readable.

File: sos/upload/targets/redhat.py (hunk 16 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py]: The idea of rounding up is to make sure we cover partial chunks, but yes I can add a note explaining this

File: sos/upload/targets/redhat.py (hunk 18 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py:207]: yes, that's right. From the specification they gave us, this is expected

**Comment by jcastill** [comment]
> I am mostly OK with the PR, thanks for the lengthy work.
> 
> Many comments I raised are rather subjective opinion or "what about" ideas of improvement - I dont insist on most of them.
> 
> I feel there are a few places here and there, where the code might get improved to be more robust, but I dont want to be too paranoic and polute the PR by my fears. Rather the daily use will reveal them, if there is any real wekaness.

I welcome the paranoia because that keeps me on my toes, and well, that's what reviews are for. My only concern is that I want something that works when the release day comes, and then after a couple of weeks of testing and verifying that works we can fix some of your concerns. Does that make sense?

File: sos/upload/targets/redhat.py (hunk 17 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py]: Not sure how to test this either, it's quite a corner case, but not the kind of bug we'd like to wait and see because we are talking about very big file attachments, so I'll see if I can add a check and make it more robust

File: sos/upload/targets/redhat.py [hunk 20]
```diff
@@ -96,6 +107,494 @@ def _get_upload_https_auth(self):
...
+        self._get_upload_session_details(archive=file_obj,
+                                         archive_size=archive_size,
+                                         total_chunks=total_chunks)
+
+        if not self._upload_session_details:
+            self.ui_log.error("Failed to get upload session details.")
+            return None
+
+        # Get session metadata
+        # The format of the upload session metadatais:
```
Comment (jcastill) [inline sos/upload/targets/redhat.py]: typo, change 'metadatais' to 'metadata is'

File: sos/upload/targets/redhat.py (hunk 19 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py:605]: These are two different things - the one in line 388 is what the server returns, and I added it there for information of what to expect, and then this line is just for debug. If the capitals/no-capitals is an issue, I can document in 388 that it's for information purposes only

File: sos/upload/targets/redhat.py (hunk 1 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py]: This is a good point. I can change it now or post-release in September, to make sure I can test the change in prod. Which approach do you prefer?

File: sos/upload/targets/redhat.py (hunk 7 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py:282]: I'll add a comment

File: man/en/sos-report.1 (hunk 9 above)
Comment (jcastill) [reply man/en/sos-report.1:531]: what about something like this? 

```
  .B \--upload-threads THREADS
  Number of threads to use for multipart uploads (1\-16, default: 4).
  When an archive exceeds a certain size, defined in the targets, it is
  uploaded in parallel chunks using this many threads. Higher values
  speed up the upload but consume more bandwidth and memory, as each
  thread holds a 128 MiB chunk in memory.

```

File: sos/upload/targets/redhat.py (hunk 13 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py]: it's a leftover of the old way. I'll rework this

File: sos/upload/targets/redhat.py (hunk 4 above)
Comment (pmoravec) [reply sos/upload/targets/redhat.py:189]: Sounds good to me.

Later on, I got the meaning, that the "get" is meant against the upload erver, not against the class (member). Still for a class method, the "get" is usually associated with "get me clamm member details".

File: sos/upload/targets/redhat.py (hunk 6 above)
Comment (pmoravec) [reply sos/upload/targets/redhat.py]: Hmm.. right no cons seen here, so I am in favour of the `self.base` change :)

File: sos/upload/targets/redhat.py (hunk 8 above)
Comment (pmoravec) [reply sos/upload/targets/redhat.py:742]: Ah that makes sense, to have those two info logs "only". Also the warning level was previously to warn user we are changing the host from customer portal to the SFTP server - this is not further the case (so no reason to warn the user).

Let keep it debug then.

File: sos/upload/targets/redhat.py (hunk 19 above)
Comment (pmoravec) [reply sos/upload/targets/redhat.py:605]: Ah ok, I mis-understood the strategy value is taken from here to there or vice versa. If they are not cross-copied or cross-checked, it is fine.

**Review by pmoravec — APPROVED** [review]
LGTM. We might rename the get session details method and raise less generic exception than `Exception`, but that is minor stuff.

File: man/en/sos-report.1 (hunk 9 above)
Comment (TurboTurtle) [reply man/en/sos-report.1:531]: I'm personally of the opinion this shouldn't be user-controllable. On the security side of things there's definitely room for abuse here (though, admittedly, that's a very small concern here today given the infrastructure RH in particular would have in place, more of a concern for if/when multipart gets exposed to more generic targets), but also I don't think users really gain a net-benefit from having this control.

Uploads perceived as "slow" likely wouldn't be realistically helped all that much by more threads, as they'd be more directly limited/influenced by the end user's bandwidth.

**Review by TurboTurtle — COMMENTED** [review]
One comment on the new option.

**Comment by pmoravec** [comment]
> One comment on the new option.

Im dont have strong preference in having this option-able. I think the option can help in some specific scenarios (big bandwidth and impatient customer => increase the threads, or low bandwidth => decrease them), but I agree the option would be rather misused than properly used. So no prob making the number as hard coded.

Is that the only feedback / is the PR good otherwise? (we would like to get it merged soon, cc @TurboTurtle and/or @bmr-cymru or @arif-ali ).

**Comment by jcastill** [comment]
> > One comment on the new option.
> 
> Im dont have strong preference in having this option-able. I think the option can help in some specific scenarios (big bandwidth and impatient customer => increase the threads, or low bandwidth => decrease them), but I agree the option would be rather misused than properly used. So no prob making the number as hard coded.

I don't either. I can see @TurboTurtle 's point, but I share @pmoravec 's view, and when I chose this option was because I wanted to give more options to users the way we do with the number of threads for report. But I'm happy with either way.

**Comment by pmoravec** [comment]
The review process got stalled here - could we get some review or decision if the option is worth to have / remove?

File: sos/upload/targets/redhat.py (hunk 11 above)
Comment (jcastill) [reply sos/upload/targets/redhat.py:625]: I missed this one, sorry. So, which exception are you thinking it's best? I can use RuntimeError, or maybe ConnectionError? But also, since at this point we got no response, so there's nothing to pass up

Comment (pmoravec) [reply sos/upload/targets/redhat.py]: `ConnectionError` sounds more appropriate to me. Good point we dont have any response specifics info.

Comment (jcastill) [reply sos/upload/targets/redhat.py]: I'll modify it then, thank you!

Comment (jcastill) [reply sos/upload/targets/redhat.py]: Done

**Review by pmoravec — APPROVED** [review]

**Comment by jcastill** [comment]
Moving this to draft, because we are going live this weekend, so we'll have the opportunity to test the code with the final version of the portal, and if we find any issues I'll be able to update this PR and not have to open a new one.

**Comment by jcastill** [comment]
Back from draft. Confirmed that the code works as expected in the new portal.

**Review by pmoravec — APPROVED** [review]
